api/datastore/sql/sql.go

```diff
@@ -151,23 +151,43 @@ func newDS(ctx context.Context, url *url.URL) (*sqlStore, error) {
 	db.SetMaxIdleConns(maxIdleConns)
 	log.WithFields(logrus.Fields{"max_idle_connections": maxIdleConns, "datastore": driver}).Info("datastore dialed")
 
-	sdb := &sqlStore{db: db}
-
-	err = sdb.runMigrations(ctx, checkExistence(db), migrations.Migrations)
-	if err != nil {
-		log.WithError(err).Error("error running migrations")
-		return nil, err
-	}
-
-	switch driver {
+	switch driver { // NOTE: fixes weird sqlite3 behavior
 	case "sqlite3":
 		db.SetMaxOpenConns(1)
 	}
-	for _, v := range tables {
-		_, err = db.ExecContext(ctx, v)
+
+	sdb := &sqlStore{db: db}
+
+	// NOTE: runMigrations happens before we create all the tables, so that it
+	// can detect whether the db did not exist and insert the latest version of
+	// the migrations BEFORE the tables are created (it uses table info to
+	// determine that).
+	//
+	// we either create all the tables with the latest version of the schema,
+	// insert the latest version to the migration table and bail without running
+	// any migrations.
+	// OR
+	// run all migrations necessary to get up to the latest, inserting that version,
+	// [and the tables exist so CREATE IF NOT EXIST guards us when we run the create queries].
```

Comment:
mostly did this for adding this note about why things are the way they are, but thought up a dreamy bug while I was in here. 11/10 level nits appreciated, my brain is a strange place that's hard to make sense of.

Reply:
Is pg failing even with 'if not exists' guards in tests?

Reply:
oh. i made a dumb dumb. yay tests

Reply:
pg is not my friend... down the rabbit hole

Reply:
found the issue https://github.com/fnproject/fn/pull/939/commits/3fba04989d9a6785fcfa5fb53d694820195ff805

Reply:
lgtm